Gulpfile.ts

```diff
@@ -0,0 +1,104 @@
+import { exec } from 'child_process';
+import { readdirSync, unlinkSync, writeFileSync } from 'fs';
+import * as gulp from 'gulp';
+import * as ts from 'gulp-typescript';
+import { basename } from 'path';
+import { promisify } from 'util';
+
+import tsConfig = require('./tsconfig.json');
+const run = promisify(exec);
+
+gulp.task('typedoc', async () => {
+  function generateTypeDocConfig() {
+    const tutorials = readdirSync('docs/reference/content/tutorials')
+      .filter(filename => filename.endsWith('.md'))
+      .map(filename => `docs/reference/content/tutorials/${filename}`);
+
+    const docOptions = {
+      entryPoint: 'types/mongodb.d.ts',
+      mode: 'file',
+      out: 'docs/gen',
+      theme: 'pages-plugin',
+      excludeNotExported: true,
+      stripInternal: true,
+      pages: {
+        enableSearch: true,
+        listInvalidSymbolLinks: true,
+        output: 'pages',
+        groups: [
+          {
+            title: 'Documentation',
+            pages: [
+              {
+                title: 'FAQ',
+                source: 'docs/reference/content/reference/faq/index.md'
+              }
+            ]
+          },
+          {
+            title: 'Tutorials',
+            pages: [
+              {
+                title: 'Quick Start',
+                source: 'docs/reference/content/quick-start/quick-start.md',
+                children: tutorials.map(filepath => ({
+                  title: basename(filepath).replace('.md', ''),
+                  source: filepath
+                }))
+              }
+            ]
+          }
+        ]
+      }
+    };
+
+    return docOptions;
+  }
+
+  const docOptions = generateTypeDocConfig();
+
+  writeFileSync('./typedoc.json', JSON.stringify(docOptions, undefined, 2), {
+    encoding: 'utf8'
+  });
+
+  try {
+    await run('npx typedoc');
```

Comment:
we shouldn't have to run `npx` in the gulp file, there are likely plugins for all of these things or at least programmatic apis. For using `typedoc` you can just use [gulp-typedoc](https://www.npmjs.com/package/gulp-typedoc) which will allow you to do all of this in a more "gulpy" way without having to write a temporary file and unlink it on error

Reply:
The gulp typedoc plugin didn't work for our project configuration, its not the plugins fault, I wrestled with it for quite a bit, the problem is actually that typedoc's programatic API differs from its config file API so this was my hacky solution, I do still have some clean up I can do here to make this more readable.